File: backend/app/schemas/payment.py
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, ConfigDict, Field

from app.models.payment import PaymentStatus


class CreatePaymentOrderRequest(BaseModel):
    order_id: int = Field(..., description="ID of the order to create payment for")


class CreatePaymentOrderResponse(BaseModel):
    order_id: int
    razorpay_order_id: str
    amount: Decimal
    amount_in_paise: int
    currency: str = "INR"
    key_id: str


class VerifyPaymentRequest(BaseModel):
    razorpay_order_id: str = Field(..., description="Razorpay Order ID")
    razorpay_payment_id: str = Field(..., description="Razorpay Payment ID")
    razorpay_signature: str = Field(..., description="Cryptographic signature from Razorpay")


class PaymentResponse(BaseModel):
    id: int
    order_id: int
    razorpay_order_id: str
    razorpay_payment_id: str | None
    amount: Decimal
    status: PaymentStatus
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/types/payment.ts
export type PaymentStatus = "PENDING" | "SUCCESS" | "FAILED" | "REFUNDED";

export interface CreatePaymentOrderResponse {
  order_id: number;
  razorpay_order_id: string;
  amount: number | string;
  amount_in_paise: number;
  currency: string;
  key_id: string;
}

export interface VerifyPaymentDto {
  razorpay_order_id: string;
  razorpay_payment_id: string;
  razorpay_signature: string;
}

export interface PaymentResponse {
  id: number;
  order_id: number;
  razorpay_order_id: string;
  razorpay_payment_id?: string | null;
  amount: number | string;
  status: PaymentStatus;
  created_at: string;
  updated_at: string;
}

export interface RazorpayOptions {
  key: string;
  amount: number;
  currency: string;
  name: string;
  description?: string;
  image?: string;
  order_id: string;
  handler: (response: {
    razorpay_payment_id: string;
    razorpay_order_id: string;
    razorpay_signature: string;
  }) => void;
  prefill?: {
    name?: string;
    email?: string;
    contact?: string;
  };
  notes?: Record<string, string>;
  theme?: {
    color?: string;
  };
  modal?: {
    ondismiss?: () => void;
  };
}

declare global {
  interface Window {
    Razorpay?: new (options: RazorpayOptions) => {
      open: () => void;
      on: (event: string, handler: (response: any) => void) => void;
    };
  }
}
